operator/src/lib.rs

```diff
@@ -282,8 +251,163 @@ impl SparkState {
         Ok(ReconcileFunctionAction::Continue)
     }
 
+    /// Write new status if not existing yet
+    /// Should only happen once after cluster creation
+    pub async fn check_status(&self) -> SparkReconcileResult {
+        if self.status.is_none() {
+            // no status set yet -> create new
+            let new_status = SparkClusterStatus {
+                current_version: self.spec.version.clone(),
+                target_version: self.spec.version.clone(),
+            };
+
+            self.context
+                .client
+                .merge_patch_status(&self.context.resource, &new_status)
+                .await?;
+
+            return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+        }
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+
+    /// Checks if all pods that currently belong to the cluster are up and running.
+    /// That means not "terminating" and "ready"
+    /// We always make sure that all masters are running before any workers / history-servers
+    async fn check_pods_up_and_running(&self) -> SparkReconcileResult {
+        trace!("Reconciliation: Checking if all pods are up and running");
+
+        if let Some(pod_info) = &self.pod_information {
+            // master nodes have priority
+            if !self.pods_up_and_running(pod_info.get_all_pods(Some(SparkNodeType::Master))) {
+                return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+            }
+
+            // worker / history-server are treated equally
+            if !self.pods_up_and_running(pod_info.get_all_pods(Some(SparkNodeType::Worker)))
+                || !self
+                    .pods_up_and_running(pod_info.get_all_pods(Some(SparkNodeType::HistoryServer)))
+            {
+                return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+            }
+        }
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+    /// Returns true only if any pod status is neither "terminating" or anything other than "ready"
+    fn pods_up_and_running(&self, pods: Vec<Pod>) -> bool {
```

Comment:
I like this. Would maybe make sense to move to operator-rs?
Could you add a TODO for that?

Reply:
Yes, i think every operator will need to check that at some point. The question is how? I use the vector approach because im not using the step by step create/delete of pods. In the zookeeper-operator you have sth. similar for one pod only.
We can go for 2 methods like pod(s)_up_and_running or just provide the single one and we loop ourselves where needed?